### PR TITLE
Correctly access the communicator name is MSGQ.

### LINKED ISSUE #12063: [5.0.0] MQD library gives corrupted communicator name
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
5.0.0

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
git clone

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.
 57c405c52ad76bab0be9f95e29a6df660673081e 3rd-party/openpmix (v4.2.7)
 1552e36f0852bbc6d901ec95983369f0a3c283f6 3rd-party/prrte (v3.0.2)
 c1cfc910d92af43f8c27807a9a84c9c13f4fbc65 config/oac (heads/main)

### Please describe the system on which you are running

* Operating system/version: Ubuntu 22.04
* Computer hardware: x86_64

-----------------------------

## Details of the problem

Communicator names retrieved via the Message Queue Debugging library libompi_dbg_msgq.so appear corrupted, for example "0��^A" instead of "MPI_COMM_WORLD".
This appears to be because in 5.0.x the c_name field in ompi_communicator_t is a char*, whereas previously it was char[MPI_MAX_OBJECT_NAME], and rebuild_communicator_list() in ompi_msgq_dll.c still assumes that the string content can be obtained as a field/segment of the communicator structure.
The rebuild_communicator_list() function would instead need to fetch the c_name field as a pointer and use it to obtain the null-terminated string.
The mqs_ interface does not provide for fetching null-terminated strings, so options would be to fetch (via the debugger) a chunk of memory which is an upper bound on the length of the string, or to (inefficiently) fetch each byte of the string until the null terminator is found.
An illustrative patch for the first is as follows:
```
diff --git a/ompi/debuggers/ompi_msgq_dll.c b/ompi/debuggers/ompi_msgq_dll.c
index 4516b8df23..b79fa5886f 100644
--- a/ompi/debuggers/ompi_msgq_dll.c
+++ b/ompi/debuggers/ompi_msgq_dll.c
@@ -682,8 +682,11 @@ static int rebuild_communicator_list (mqs_process *proc)
                                     p_info );
             old->group = find_or_create_group( proc, group_base );
         }
-        mqs_fetch_data( proc, comm_ptr + i_info->ompi_communicator_t.offset.c_name,
-                        64, old->comm_info.name );
+        mqs_taddr_t name_addr = ompi_fetch_pointer( proc, comm_ptr + i_info->ompi_communicator_t.offset.c_name, p_info );
+        mqs_fetch_data( proc, name_addr, 64, old->comm_info.name );
+        old->comm_info.name[63] = '\0';
+        size_t name_len = strlen(old->comm_info.name);
+        memset(&old->comm_info.name[name_len], 0, 64-name_len);
 
         if( NULL != old->group ) {
             old->comm_info.size = old->group->entries;
```

### PR DESCRIPTION
**Pull request by bosilca**
We transformed the communicator c_name field from an array into a pointer and fialed to update the MSGQ accessors to the field.

Thanks @david-edwards-linaro for the report and the initial patch.

Fixes #12063.

### CODE REVIEW

**Comment by jsquyres** [comment]
@davideberius Could you test this patch (on Open MPI's `main` branch) and see if it fixes the problem for you?

**Comment by david-edwards-linaro** [comment]
I ran into some prrte build issues on main, though I applied this change to 5.0.0 and was able to see the communicator names as expected.
There was some additional content after the first null in the mqs_communicator char[64] array (e.g. MPI_COMM_SELF\0SCRIBE\0...) which could be cleaned up as you prefer (we're fetching 64 bytes, but the name may be shorter than this, so we also fetch adjacent data not relating to the name).

**Comment by jsquyres** [comment]
Ah, we didn't realize / remember that it shows all 64 bytes, not just up to the `\0` termination.

@bosilca Could you adjust?

**Comment by bosilca** [comment]
That is ugly ! But, this code is outside the performance critical code, I will make the adjustements.

**Comment by bosilca** [comment]
@david-edwards-linaro I have pushed the update to zero the remainder of the name buffer. If you can please give it a try.

**Comment by david-edwards-linaro** [comment]
Thanks for the update.  I'm still seeing some non-null content - looking through the code, a couple of things came to mind:
- the second and third arguments to memset appear reversed (character then count rather vs. count then character)
- I've not tested a communicator name length of 63 (or more), but I wonder if there may be an issue with giving memset a count of (62 - strlen(name)) in such a scenario

**Comment by jsquyres** [comment]
@bosilca I don't think that this patch is right.  Per https://github.com/open-mpi/ompi/blob/4984355351583ad7b9a23f745c3923ff64434e9b/ompi/communicator/comm.c#L2014-L2023, don't we have to look at `communicator.c_flags` and see if `OMPI_COMM_NAMEISSET` is set?

I'm actually unclear on why we need to memset beyond the end of the name in the target buffer.  The target buffer should **always** contain a `\0`; is the debugger showing information beyond this `\0`?

**Comment by david-edwards-linaro** [comment]
With this approach the debugger will fetch 64 bytes from the target's address space even if the name is shorter - it's optional whether you choose to zero this additional data out or rely on the debugger displaying data to the user only up to the first null.

**Comment by jsquyres** [comment]
> rely on the debugger displaying data to the user only up to the first null.

Yes, that's exactly what I am assuming.  Is this not happening?  I.e., is the debugger displaying the entire 64 bytes, regardless of whether there's a `\0` in the middle?

I also wonder if something else is going on, because @bosilca's latest patch *should* zero out all the rest of the name, but you reported that you're still seeing unrelated characters in the debugger's output.

**Comment by david-edwards-linaro** [comment]
In our case the debugger shows data only up to the first null... that's not quite the same as mandating all debuggers do this.  If it's deemed unnecessary to zero out the additional data, just fixing the primary issue (i.e. fetch the pointer then fetch the data at that address) will be fine.

**Comment by jsquyres** [comment]
> In our case the debugger shows data only up to the first null... that's not quite the same as mandating all debuggers do this. If it's deemed unnecessary to zero out the additional data, just fixing the primary issue (i.e. fetch the pointer then fetch the data at that address) will be fine.

Ah, ok good.  So your comment:

> I'm still seeing some non-null content 

Can probably be attributed to the fact that the arguments to memset are backwards.  Let me fix that, and let's see if that solves the issue.  I'll still leave in the "let's zero out everything else", just as a good defensive programming measure.

**Comment by jsquyres** [comment]
@bosilca I pushed two commits:

1. One intended to be squashed into yours, if you agree with the changes.
2. A separate one to fix a comment and remove a debug `printf()`.

Can you review?

I think we still need to fix the `OMPI_COMM_NAMEISSET`, but I'm out of time at the moment.

File: ompi/debuggers/ompi_msgq_dll.c [hunk 1]
```diff
@@ -682,8 +684,21 @@ static int rebuild_communicator_list (mqs_process *proc)
...
+                                                    p_info );
+        /* c_name can be up to MPI_MAX_OBJECT_NAME bytes, but we only
+         * copy the first (sizeof(old->comm_info.name)-1) here.  Make
+         * sure the string is correctly terminated. */
+        mqs_fetch_data( proc, name_addr, 64, old->comm_info.name );
+        size_t target_size = sizeof(old->comm_info.name);
+        old->comm_info.name[target_size - 1] = '\0';
+        /* Defensively zero anything beyond the actual name */
+        size_t src_strlen = strlen(old->comm_info.name);
+        if (target_size - 1 > src_strlen) {
```
Comment (bosilca) [inline ompi/debuggers/ompi_msgq_dll.c:698]: This condition can never happen, the line above sets the termination at position `target_size - 1` so the strlen cannot return anything larger than that. When they are equal memset will not do anything, so this test is unnecessary.

**Comment by jsquyres** [comment]
@david-edwards-linaro Could you try again with the current patches?  I think we still have one more minor thing to fix, but getting confirmation that the name is coming across properly would be great.

**Comment by david-edwards-linaro** [comment]
I've applied `877ee24` and `ed4b78d` to v5.0.x (`2c4740d`) and see MPI_COMM_WORLD, MPI_COMM_SELF and MPI_COMM_NULL as expected.

**Comment by jsquyres** [comment]
Thanks for the confirmation, @david-edwards-linaro.

@bosilca Could you amend this PR to check for the `OMPI_COMM_NAMEISSET` flag?

**Comment by bosilca** [comment]
I don't think it is necessary. We always put a name when we create communicators to indicate their history, and the user will overwrite it via `MPI_Comm_set_name`. Thus, the user would always benefit from having access to this name.

**Comment by jsquyres** [comment]
> I don't think it is necessary to check if the user set a name. OMPI always set a default name on communicator creation to indicate their history, and the user will overwrite it via `MPI_Comm_set_name`. Thus, users would always benefit from having access to this name.

Ah, gotcha!  So I guess we're done.

**Review by jsquyres — APPROVED** [review]

**Pull request merged**